File: TestLeetCode/TreeNodeDistance.cs
/**
 * Definition of TreeNode:
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left, right;
 *     public TreeNode(int val) {
 *         this.val = val;
 *         this.left = this.right = null;
 *     }
 * }
 */

namespace lintcode
{
    class Solution {
        /**
         * @param root: The root node of a binary tree.
         * @param p: The value of a node in the binary tree.
         * @param q: The value of a node in the binary tree.
         * @return: The distance between two nodes in a binary tree.
         */
        public int CalculateDistance(TreeNode root, int p, int q) {
            var lca = GetLCA(root, p, q);

            int distToP = CalDist(lca, p, 0);
            int distToQ = CalDist(lca, q, 0);
            return distToP + distToQ;
        }

        private int CalDist(TreeNode root, int p, int dist) {
            if (root == null) {
                return Int32.MaxValue;
            }
            if (root.val == p) {
                return dist;
            }

            return Math.Min(CalDist(root.left, p, dist+1), CalDist(root.right, p, dist+1));
        }

        /// Get least common acesstors
        private TreeNode GetLCA(TreeNode root, int p, int q) {
            if (root == null) {return null;}

            if (root.val == p || root.val == q) {
                return root;
            }
            var left = GetLCA(root.left, p, q);
            var right = GetLCA(root.right, p, q);

            if (left != null && right != null) {
                return root;
            }
            return left ?? right;
        }
    }
}

File: python/tree/tree_node_distance.py
from python.common import TreeNode


class TreeNodeDistance:
    def calculate_distance(self, root: TreeNode | None, left: int, right: int) -> int:
        def path(node, target, distance=0):
            if not node: return None
            if node.val == target: return distance
            found = path(node.left, target, distance + 1)
            return found if found is not None else path(node.right, target, distance + 1)
        def lca(node):
            if not node or node.val in {left, right}: return node
            a, b = lca(node.left), lca(node.right)
            return node if a and b else a or b
        ancestor = lca(root)
        if not ancestor: raise ValueError("nodes not found")
        first, second = path(ancestor, left), path(ancestor, right)
        if first is None or second is None: raise ValueError("nodes not found")
        return first + second
